README.md
# Sitemap generator for SilverStripe 3
This extension adds sitemap functionality to your SilverStripe website. It is heavily based on
[googlesitemaps](https://github.com/silverstripe-labs/silverstripe-googlesitemaps), however does things a little
differently, adding in support for basic custom filters when adding DataObjects, and removing some CMS integration,
pagination of results, and JavaScript support. The sitemap is hard-coded to have a maximum of 1000 dataobjects per
sitemap (due to memory use), and will automatically split your sitemaps into multiple if there are more.

## Requirements
* SilverStripe 3.0+

## Installation
```
composer require axllent/silverstripe-sitemap
```

## Basic usage
Simply install the module. To check the sitemap is working, go to
`http://yoursite.com/sitemap.xml?flush=1`. By default it will add your SiteTree pages (with "show in search" enabled),
however you can also optionally add DataObjects provided they are mapped to URLs.

## Specifying the Sitemap location in your robots.txt file
```
Sitemap: http://yoursite.com/sitemap.xml
```

## Adding DataObjects
DataObjects can easily be added too, assuming of course that those DataObjects have unique URLs on your site.
DatObjects must either have a `Link()` function, or alternatively have a `SitemapAbsoluteURL()` function (which will
override `Link()`). The `SitemapAbsoluteURL()` function must return an absolute URL, `Link()` will get automatically
converted.

To add a DataObject to your sitemap you can simply add the following to your `mysite/_config.php`:
```php
Sitemap::register_dataobject('MyDataObject');
```

Additional options are available:
```php
Sitemap::register_dataobject('MyDataObject', array(
	'filter' => array('ShowOnWeb' => 1),
	'where' => '"ExpiryDate" >= \'' . date('Y-m-d') . '\' OR "ExpiryDate" IS NULL',
	'exclude' => array('StockLevel' => 0),
	'filterByCallback' => function($dataobject, $list) {return $dataobject->canView();}
	'frequency' => 'weekly', // always, hourly, daily, weekly, monthly, yearly, never
	'priority' => 0.5 // Valid values range from 0.0 to 1.0
));
```

## Google Notifications
Publishing & unpublishing of SiteTree pages can automatically send a "ping" to Google. Please note
that this is turned off by default and does not apply to saving of DataObjects. Also note that you must have
registered your sitemap with Google's webmaster tools first.

To enable change notifications to Google:

```php
Sitemap::enable_google_notifications();
```


CHANGELOG.md
# Changelog

Notable changes to this project will be documented in this file.

## [1.0.2]

- Return 404 on invalid request


## [1.0.1]

- Fix SilverStripe framework version requirement


## [1.0.0]

- Adopt semantic versioning releases


code/extensions/SitemapExtension.php
<?php
/**
 * Decorate the page object to provide sitemaps
 *
 * Based on github.com/silverstripe-labs/silverstripe-googlesitemaps
 */

class SitemapExtension extends DataExtension
{

    /**
     * The default value of the priority field depends on the depth of the page in
     * the site tree, so it must be calculated dynamically.
     *
     * @return mixed
     */
    public function getGooglePriority()
    {
        return Sitemap::get_priority_for_class($this->owner->class);
    }

    /**
     * Returns a pages change frequency calculated by pages age and number of
     * versions. Google expects always, hourly, daily, weekly, monthly, yearly
     * or never as values.
     *
     * @see http://support.google.com/webmasters/bin/answer.py?hl=en&answer=183668&topic=8476&ctx=topic
     *
     * @return SS_Datetime
     */
    public function getChangeFrequency()
    {
        if ($freq = Sitemap::get_frequency_for_class($this->owner->class)) {
            return $freq;
        }

        $date = date('Y-m-d H:i:s');

        $created = new SS_Datetime();
        $created->value = ($this->owner->Created) ? $this->owner->Created : $date;

        $now = new SS_Datetime();
        $now->value = $date;

        $versions = ($this->owner->Version) ? $this->owner->Version : 1;
        $timediff = $now->format('U') - $created->format('U');

        // Check how many revisions have been made over the lifetime of the
        // Page for a rough estimate of it's changing frequency.
        $period = $timediff / ($versions + 1);

        if ($period > 60 * 60 * 24 * 365) {
            $freq = 'yearly';
        } elseif ($period > 60 * 60 * 24 * 30) {
            $freq = 'monthly';
        } elseif ($period > 60 * 60 * 24 * 7) {
            $freq = 'weekly';
        } elseif ($period > 60 * 60 * 24) {
            $freq = 'daily';
        } elseif ($period > 60 * 60) {
            $freq = 'hourly';
        } else {
            $freq = 'always';
        }

        return $freq;
    }

    /*
     * @return void
     */
    public function onAfterPublish()
    {
        Sitemap::ping();
    }

    /*
     * @return void
     */
    public function onAfterUnpublish()
    {
        Sitemap::ping();
    }
}


code/controllers/Sitemap_Controller.php
<?php

/**
 * Controller for displaying the sitemap.xml. The module displays an index
 * sitemap at the sitemap.xml level, then outputs the individual objects
 * at a second level.
 *
 * <code>
 * http://site.com/sitemap.xml
 * http://site.com/sitemap.xml/sitemap/$ClassName
 * </code>
 *
 * Originally based on github.com/silverstripe-labs/silverstripe-googlesitemaps
 */
class Sitemap_Controller extends Controller
{

    private static $allowed_actions = array(
        'index',
        'sitemap'
    );

    public static $url_handlers = array(
        'sitemap/$Class/$PageNum' => 'sitemap'
    );

    /**
     * Default controller action for the sitemap.xml file. Renders a index
     * file containing a list of links to sub sitemaps containing the data.
     */
    public function index($url)
    {
        $this->Sitemaps = Sitemap::get_sitemaps();

        if ($this->Sitemaps->Count() == 0) {
            return new SS_HTTPResponse('Page not found', 404);
        }

        foreach ($this->Sitemaps as $s) {
            $s->SitemapAbsoluteURL = Director::absoluteURL('sitemap.xml/sitemap/' . $s->ClassName . '/');
        }

        $this->getResponse()->addHeader('Content-Type', 'application/xml; charset="utf-8"');
        $this->getResponse()->addHeader('X-Robots-Tag', 'noindex');

        return $this->renderWith('Sitemap_sitemaps');
    }

    /**
     * Specific controller action for displaying a particular list of links
     * for a class
     */
    public function sitemap($request)
    {
        $class = $this->request->param('Class');
        $page_num = $this->request->param('PageNum');

        $page_num = (is_numeric($page_num) && $page_num > 0) ? $page_num : 1;

        $this->Items = Sitemap::get_items($class, $page_num);

        if (!$this->Items || $this->Items->Count() == 0) {
            return new SS_HTTPResponse('Page not found', 404);
        }

        $this->getResponse()->addHeader('Content-Type', 'application/xml; charset="utf-8"');
        $this->getResponse()->addHeader('X-Robots-Tag', 'noindex');

        return $this->renderWith('Sitemap');
    }
}


code/extensions/SitemapSiteTreeExtension.php
<?php
/**
 * SiteTree extension to calculate the update "Google Priority"
 *
 * Originally based on github.com/silverstripe-labs/silverstripe-googlesitemaps
 */

class SitemapSiteTreeExtension extends SitemapExtension
{

    /**
     * Ensure that all parent pages of this page (if any) are published
     * @return boolean
     */
    public function hasPublishedParent()
    {
        // Skip root pages
        if (empty($this->owner->ParentID)) {
            return true;
        }

        // Ensure direct parent exists
        $parent = $this->owner->Parent();
        if (empty($parent) || !$parent->exists()) {
            return false;
        }

        // Check ancestry
        return $parent->hasPublishedParent();
    }

    /**
     * @return mixed
     */
    public function getGooglePriority()
    {
        setlocale(LC_ALL, "en_US.UTF8");

        $parentStack = $this->owner->parentStack();
        $numParents = is_array($parentStack) ? count($parentStack) - 1 : 0;

        $num = max(0.1, 1.0 - ($numParents / 10));
        $result = str_replace(",", ".", $num);

        return $result;
    }
}


code/Sitemap.php
<?php
/**
 * Sitemap support for SilverStripe
 * ================================
 *
 * Extension to sitemap support SilverStripe
 *
 * Usage: See README.md
 *
 * License: MIT-style license http://opensource.org/licenses/MIT
 * Authors: Techno Joy development team (www.technojoy.co.nz)
 *
 * Based on github.com/silverstripe-labs/silverstripe-googlesitemaps
 */

class Sitemap
{

    private static $include_sitetree = true;

    private static $google_notifications = false;

    private static $dataobjects = array();

    public static function register_dataobject($className, $options = array())
    {
        if (!$className || !is_array($options)) {
            trigger_error('SimpleSitemap::register_dataobject($className, $options=array())', E_USER_ERROR);
            return false;
        }
        if (!self::is_registered($className) && class_exists($className)) {
            $className::add_extension('SitemapExtension');
            self::$dataobjects[$className] = $options;
        }
    }

    /**
     * Do not include SiteTree
     *
     * @param null
     * @return null
     */
    public static function exclude_sitetree()
    {
        self::$include_sitetree = false;
    }

    /**
     * Enable Google pings
     *
     * @param null
     * @return null
     */
    public static function enable_google_notifications()
    {
        self::$google_notifications = false;
    }

    /**
     * Is class already registered?
     *
     * @param String
     * @return boolean
     */
    public static function is_registered($className)
    {
        return isset(self::$dataobjects[$className]);
    }

    /**
     * Remove class from list
     *
     * @param String
     * @return null
     */
    public static function unregister_dataobject($className)
    {
        if (self::is_registered($className)) {
            unset(self::$dataobjects[$className]);
        }
    }

    /**
     * Returns the string frequency of edits for a particular dataobject class
     * Frequency for {@link SiteTree} objects can be determined from the version history.
     *
     * @param string
     * @return string
     */
    public static function get_frequency_for_class($class)
    {
        if (self::is_registered($class) && isset(self::$dataobjects[$class]['frequency'])) {
            return self::$dataobjects[$class]['frequency'];
        }
    }

    /**
     * Returns the default priority of edits for a particular dataobject class
     *
     * @param string
     * @return float
     */
    public static function get_priority_for_class($class)
    {
        if (self::is_registered($class) && isset(self::$dataobjects[$class]['priority'])) {
            return self::$dataobjects[$class]['priority'];
        }
        return 0.5;
    }

    /**
     * Automatically add SiteTree
     *
     * @param null
     * @return null
     */
    public static function init_sitetree()
    {
        if (self::$include_sitetree && class_exists('SiteTree') && !self::is_registered('SiteTree')) {
            self::register_dataobject('SiteTree', array(
                'filter' => array('ShowInSearch' => 1),
                'filterByCallback' => function ($page, $list) {
                    return $page->canView();
                }
            ));
        }
    }

    /**
     * Return all sitemaps
     *
     * @param null
     * @return ArrayList
     */
    public static function get_sitemaps()
    {
        self::init_sitetree();

        $sitemaps = new ArrayList();

        if (count(self::$dataobjects) > 0) {
            foreach (self::$dataobjects as $class => $config) {
                $items = self::get_filtered_results($class);

                if ($items->Count() > 0) {
                    $list = new PaginatedList($items);
                    $list->setPageLength(1000);
                    $pages = $list->TotalPages();

                    for ($x=1; $x <= $pages; $x++) {
                        $latest = $items->limit(1000, ($x-1) * 1000)->Sort('LastEdited', 'DESC')->first();
                        $sitemaps->push(new ArrayData(array(
                            'ClassName' => $class,
                            'LastEdited' => date('Y-m-d', strtotime($latest->LastEdited)),
                            'Page' => ($x == 1) ? false : $x
                        )));
                    }
                }
            }
        }

        return $sitemaps;
    }

    /**
     * Return all filtered items in class
     * Supports additional filtering (filter, where & exclude)
     *
     * @param string
     * @return ArrayList
     */
    public static function get_filtered_results($className)
    {
        self::init_sitetree();

        if (!self::is_registered($className)) {
            return false;
        }

        $class = self::$dataobjects[$className];

        $list = new DataList($className);
        if (isset($class['filter'])) {
            $list = $list->filter($class['filter']);
        }
        if (isset($class['where'])) {
            $list = $list->where($class['where']);
        }
        if (isset($class['exclude'])) {
            $list = $list->exclude($class['exclude']);
        }
        if (isset($class['filterByCallback'])) {
            $list = $list->filterByCallback($class['filterByCallback']);
        }
        $list = $list->sort('LastEdited', 'DESC');
        return $list;
    }

    /**
     * Return paginated results in class
     * Only returns items that contain Link()
     * Limited to 1000 results per page
     *
     * @param string
     * @return ArrayList
     */
    public static function get_items($className, $page=1)
    {
        $items = self::get_filtered_results($className);
        if (!$items) {
            return new ArrayList(); // 404
        }
        $list = new PaginatedList($items);
        $list->setPageLength(1000);
        $list->setCurrentPage($page);

        $output = new ArrayList();

        /* only push items with a link */
        foreach ($list as $item) {
            $item->ChangeFrequency = self::get_frequency_for_class($className);
            $item->GooglePriority = self::get_priority_for_class($className);
            if ($item->hasMethod('SitemapAbsoluteURL')) {
                $item->SitemapAbsoluteURL = $SitemapAbsoluteURL->SitemapAbsoluteURL();
                $output->push($item);
            } elseif ($item->hasMethod('Link')) {
                $item->SitemapAbsoluteURL = Director::absoluteURL($item->Link());
                $output->push($item);
            }
        }

        /* Make sure we only include one of each link, and no external links (ie: redirector pages */
        $output->removeDuplicates('SitemapAbsoluteURL');
        $external_links = array();
        $base_url = preg_quote(Director::absoluteBaseURL(), '/');
        foreach ($output as $item) {
            if (!preg_match('/^' . $base_url . '/', $item->SitemapAbsoluteURL)) {
                array_push($external_links, $item->SitemapAbsoluteURL);
            }
        }
        if (count($external_links) > 0) {
            $output = $output->exclude('SitemapAbsoluteURL', $external_links);
        }

        return $output;
    }

    /**
     * Notifies Google about changes to your sitemap. This behavior is disabled
     * by default, to enable, read the documentation provided in the docs folder.
     *
     * After notifications have been enabled, every publish / unpublish of a page.
     * will notify Google of the update.
     *
     * If the site is in development mode no ping will be sent regardless whether
     * the Google notification is enabled.
     *
     * @return string Response text
     */
    public static function ping()
    {
        if (!self::$google_notifications) {
            return false;
        }

        if (!Director::isLive()) {
            return;
        }

        $location = urlencode(Controller::join_links(
            Director::absoluteBaseURL(),
            'sitemap.xml'
        ));

        $response = self::send_ping(
            'www.google.com', '/webmasters/sitemaps/ping', sprintf('sitemap=%s', $location)
        );

        return $response;
    }

    /**
     * Send an HTTP request to the host.
     *
     * @return String Response text
     */
    protected static function send_ping($host, $path, $query)
    {
        $socket = fsockopen($host, 80, $errno, $error);
        if (!$socket) {
            return $error;
        }
        if ($query) {
            $query = '?' . $query;
        }
        $request = "GET {$path}{$query} HTTP/1.1\r\nHost: $host\r\nConnection: Close\r\n\r\n";
        fwrite($socket, $request);
        $response = stream_get_contents($socket);

        return $response;
    }
}
